core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
+    /** Returns true if the source has a known finite length, false otherwise. */
+    boolean hasLength();
+
+    /** Returns the length if {@link #hasLength()} is true, otherwise throws an exception. */
+    default long length() { throw new UnsupportedOperationException(); }
+
+    /** Returns the offset of the next byte that will be read by any operation. */
+    long position();
+
+    /** Returns true if the source can skip bytes without reading them. */
+    boolean canNavigateForward();
+
+    /** Returns true if the source can always be repositioned behind the current position. Note that a buffered
+     * source may allow limited backwards navigation within the buffered data, while returning false here. */
+    boolean canNavigateBackwards();
+
+    /** Position the source at numBytes after the current position. May be implemented by skipping if the
+     * underlying source is capable of skipping, or by reading and discarding numBytes.
+     *
+     * @param numBytes the number of bytes desired to be skipped.
+     * @return either numBytes or a smaller number if EOF is hit before numBytes is read
+     */
+    int navigateForward(final long numBytes); // TODO: implement default method that reads
+
+    /** Position the source at numBytes before the current position.
+     *
+     * @param numBytes the number of bytes desired to be skipped.
+     * @return either numBytes or a smaller number if numBytes > position
+     */
+    default int navigateBackward(final long numBytes) {
+        throw new UnsupportedOperationException();
+    }
+
+    /** Positions the source such that the next byte read will be at offset `position`. */
+    default int navigateToPosition(final long position) {
+        throw new UnsupportedOperationException();
+    }
+
```

Comment:
We should have an EOF() method of some sort as a standard way to tell if we're at the end of file/stream.  Otherwise people will implement it inconsistently themselves.

Reply:
Sure.